CSharp/Library/Microsoft.Bot.Builder/Dialogs/PromptRecognizers.cs

```diff
@@ -0,0 +1,476 @@
+        /// <param name="message">Message context.</param>
+        /// <param name="expressionKey">Name of the resource with the RegEx.</param>
+        /// <param name="resourceManager">Resources with the localized expression.</param>
+        IEnumerable<RecognizeEntity<string>> RecognizeLocalizedRegExp(IMessageActivity message, string expressionKey, ResourceManager resourceManager);
+
+        /// <summary>Recognizer for a number.</summary>
+        /// <param name="message">Message context.</param>
+        /// <param name="synonymsDictionary">Dictionary with the options to choose from as a key and their synonyms as a value.</param>
+        /// <param name="options">Options of the Recognizer. <see cref="IPromptRecognizeChoicesOptions" /></param>
+        IEnumerable<RecognizeEntity<T>> RecognizeChoices<T>(IMessageActivity message, IDictionary<T, IEnumerable<T>> synonymsDictionary, IPromptRecognizeChoicesOptions options = null);
+
+        /// <summary>Recognizer for a number.</summary>
+        /// <param name="message">Message context.</param>
+        /// <param name="synonymsKey">Name of the resource with the synonyms.</param>
+        /// <param name="resourceManager">Resources with the localized synonyms.</param>
+        /// <param name="options">Options of the Recognizer. <see cref="IPromptRecognizeChoicesOptions" /></param>
+        IEnumerable<RecognizeEntity<string>> RecognizeLocalizedChoices(IMessageActivity message, string synonymsKey, ResourceManager resourceManager, IPromptRecognizeChoicesOptions options = null);
+
+        /// <summary>Recognizer for a number.</summary>
+        /// <param name="message">Message context.</param>
+        /// <param name="options">Options of the Recognizer. <see cref="IPromptRecognizeNumbersOptions" /></param>
+        IEnumerable<RecognizeEntity<double>> RecognizeNumbers(IMessageActivity message, IPromptRecognizeNumbersOptions options = null);
+
+        /// <summary>Recognizer for a ordinal number.</summary>
+        /// <param name="message">Message context.</param>
+        IEnumerable<RecognizeEntity<long>> RecognizeOrdinals(IMessageActivity message);
+
+        /// <summary>Recognizer for a time or duration.</summary>
+        /// <param name="message">Message context.</param>
+        IEnumerable<RecognizeEntity<string>> RecognizeTimes(IMessageActivity message);
+
+        /// <summary>Recognizer for true/false expression.</summary>
+        /// <param name="message">Message context.</param>
+        IEnumerable<RecognizeEntity<bool>> RecognizeBooleans(IMessageActivity message);
+    }
+
+    internal class SynonymsDictionary : Dictionary<string, IEnumerable<string>> { }
+
+    internal class LocalizedDictionary<T> : ConcurrentDictionary<string, T> { }
+
+    internal class ResourcesCache<T> : ConcurrentDictionary<string, LocalizedDictionary<T>> { }
+
+    [Serializable]
+    public class PromptRecognizers : IPromptRecognizers
```

Comment:
>IPromptRecognizers [](start = 37, length = 18)

probably PromptRecognizer and IPromptRecognizer (both singular) are better names.